src/libraries/TwabLib.sol

```diff
@@ -18,435 +17,674 @@ import { ObservationLib, MAX_CARDINALITY } from "./ObservationLib.sol";
+        timestamp: currentTime
+      });
+
+      // Write to storage
+      _account.observations[index] = observation;
+    }
+    // Write to storage
+    _account.details = accountDetails;
   }
 
   /**
-   * @notice Calculates the next TWAB checkpoint for an account with a decreasing delegateBalance.
-   * @dev    With Account struct and amount decreasing calculates the next TWAB observable checkpoint.
-   * @param _account        Account whose delegateBalance will be decreased
-   * @param _amount         Amount to decrease the delegateBalance by
-   * @param _delegateAmount The amount to increase the delegateBalance by
-   * @param _revertMessage  Revert message for insufficient delegateBalance
-   * @return accountDetails Updated AccountDetails struct
-   * @return twab           TWAB observation (with decreasing average)
-   * @return isNewTwab      Whether TWAB is new or calling twice in the same block
+   * @notice Decrease a user's balance and delegate balance by a given amount.
+   * @dev This function mutates the provided account.
+   * @param _account The account to update
+   * @param _amount The amount to decrease the balance by
+   * @param _delegateAmount The amount to decrease the delegate balance by
+   * @param _revertMessage The revert message to use if the balance is insufficient
+   * @return observation The new/updated observation
+   * @return isNew Whether or not the observation is new or overwrote a previous one
+   * @return isObservationRecorded Whether or not the observation was recorded to storage
    */
   function decreaseBalances(
     Account storage _account,
-    uint112 _amount,
-    uint112 _delegateAmount,
-    uint32 _overwritePeriod,
+    uint96 _amount,
+    uint96 _delegateAmount,
     string memory _revertMessage
   )
     internal
-    returns (
-      AccountDetails memory accountDetails,
-      ObservationLib.Observation memory twab,
-      bool isNewTwab
-    )
+    returns (ObservationLib.Observation memory observation, bool isNew, bool isObservationRecorded)
   {
-    accountDetails = _account.details;
+    AccountDetails memory accountDetails = _account.details;
 
     require(accountDetails.balance >= _amount, _revertMessage);
     require(accountDetails.delegateBalance >= _delegateAmount, _revertMessage);
 
-    if (_delegateAmount != uint112(0)) {
-      (accountDetails, twab, isNewTwab) = _nextTwab(
-        _account.twabs,
-        accountDetails,
-        _overwritePeriod
-      );
-    }
+    uint32 currentTime = uint32(block.timestamp);
+    uint32 index;
+    ObservationLib.Observation memory newestObservation;
+    isObservationRecorded = _delegateAmount != uint96(0);
 
     unchecked {
       accountDetails.balance -= _amount;
       accountDetails.delegateBalance -= _delegateAmount;
     }
+
+    // Only record a new Observation if the users delegateBalance has changed.
+    if (isObservationRecorded) {
```

Comment:
Lots of code similar to `increaseBalances`> Would be nice to have an helper function to update the observation.